File: crates/core/src/history.rs
//! The archive from Clement87/Pepito-data reduced to
//! the same `Sighting` values as the live stream, plus statistics.

use chrono::{DateTime, Datelike, Timelike};
use serde::{Deserialize, Serialize};

use crate::types::{Sighting, Way};

#[derive(Deserialize)]
struct Tweet {
  full_text: String,
  way: String,
  created_at: String,
  #[serde(default)]
  media: String,
}

/// Parses Twitter's own format.
#[must_use]
pub fn parse_twitter_time(text: &str) -> Option<i64> {
  DateTime::parse_from_str(text, "%a %b %d %H:%M:%S %z %Y")
    .ok()
    .map(|parsed| parsed.timestamp())
}

/// Parses the tweet archive into sightings, sorted by time
///
/// Entries with an unknown direction or a broken date are skipped rather than
/// failing the whole file.
///
/// # Errors
///
/// Returns the serde error when the input is not a JSON array of tweets. It
/// carries the line and column, which a stringly-typed error would throw away.
pub fn parse_history(json: &str) -> Result<Vec<Sighting>, serde_json::Error> {
  let tweets: Vec<Tweet> = serde_json::from_str(json)?;
  let mut sightings: Vec<Sighting> = tweets
    .into_iter()
    .filter_map(|tweet| {
      Some(Sighting {
        way: tweet.way.parse().ok()?,
        time: parse_twitter_time(&tweet.created_at)?,
        img: if tweet.media.is_empty() {
          None
        } else {
          Some(tweet.media)
        },
        text: Some(tweet.full_text),
      })
    })
    .collect();
  sightings.sort_by_key(|sighting| sighting.time);
  Ok(sightings)
}

#[derive(Debug, Clone, Copy, Serialize)]
pub struct Outing {
  pub out_at: i64,
  pub in_at: i64,
  pub secs: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct Stats {
  pub total: usize,
  pub ins: usize,
  pub outs: usize,
  pub first: Option<i64>,
  pub last: Option<i64>,
  pub outings: usize,
  /// Departures with no return, in other words holes in the data.
  pub unpaired: usize,
  pub avg_outing_secs: i64,
  pub median_outing_secs: i64,
  pub longest: Option<Outing>,
  pub shortest: Option<Outing>,
  pub by_hour_out: [u32; 24],
  pub by_hour_in: [u32; 24],
  /// 0 = Sunday
  pub by_weekday_out: [u32; 7],
}

/// Pairs every `in` with the last `out` before it
///
/// Two `out` events in a row (and the archive has a fair few) overwrite each
/// other, because they mean a return went missing.
#[must_use]
pub fn stats(sightings: &[Sighting]) -> Stats {
  let mut stats = Stats {
    total: sightings.len(),
    ins: 0,
    outs: 0,
    first: sightings.first().map(|sighting| sighting.time),
    last: sightings.last().map(|sighting| sighting.time),
    outings: 0,
    unpaired: 0,
    avg_outing_secs: 0,
    median_outing_secs: 0,
    longest: None,
    shortest: None,
    by_hour_out: [0; 24],
    by_hour_in: [0; 24],
    by_weekday_out: [0; 7],
  };
  let mut pending: Option<i64> = None;
  let mut durations: Vec<i64> = Vec::new();

  for sighting in sightings {
    let utc = DateTime::from_timestamp(sighting.time, 0);
    let hour = utc.map_or(0, |moment| moment.hour() as usize);
    match sighting.way {
      Way::Out => {
        stats.outs += 1;
        stats.by_hour_out[hour] += 1;
        if let Some(moment) = utc {
          stats.by_weekday_out[moment.weekday().num_days_from_sunday() as usize] += 1;
        }
        if pending.replace(sighting.time).is_some() {
          stats.unpaired += 1;
        }
      }
      Way::In => {
        stats.ins += 1;
        stats.by_hour_in[hour] += 1;
        if let Some(out_at) = pending.take() {
          let outing = Outing {
            out_at,
            in_at: sighting.time,
            secs: sighting.time - out_at,
          };
          durations.push(outing.secs);
          if stats
            .longest
            .is_none_or(|longest| outing.secs > longest.secs)
          {
            stats.longest = Some(outing);
          }
          if stats
            .shortest
            .is_none_or(|shortest| outing.secs < shortest.secs)
          {
            stats.shortest = Some(outing);
          }
        }
      }
    }
  }
  if pending.is_some() {
    stats.unpaired += 1;
  }
  stats.outings = durations.len();
  if !durations.is_empty() {
    let count = i64::try_from(durations.len()).unwrap_or(i64::MAX);
    stats.avg_outing_secs = durations.iter().sum::<i64>() / count;
    durations.sort_unstable();
    stats.median_outing_secs = durations[durations.len() / 2];
  }
  stats
}

File: php/src/History.php
<?php

declare(strict_types=1);

namespace Pepito;

use DateTimeImmutable;
use DateTimeZone;
use JsonException;
use Pepito\Exception\InvalidArchiveException;
use stdClass;

/**
 * The tweet archive (see Client::ARCHIVE_URL) reduced to the same sightings as
 * the live stream, plus statistics. Plain PHP, the default backend of Client.
 *
 * @internal Use Client::historyStats() and Client::historyParse().
 */
final class History
{
    private const MONTHS = [
        'Jan' => 1, 'Feb' => 2, 'Mar' => 3, 'Apr' => 4, 'May' => 5, 'Jun' => 6,
        'Jul' => 7, 'Aug' => 8, 'Sep' => 9, 'Oct' => 10, 'Nov' => 11, 'Dec' => 12,
    ];

    private const WEEKDAYS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

    /**
     * Sightings sorted by time. Entries with an unknown direction or a broken
     * date are skipped; a file that is not an array of tweets throws.
     *
     * @return list<array{way: string, time: int, img?: string, text?: string}>
     */
    public static function parse(string $json): array
    {
        try {
            // Objects stay objects, so that `{}` is not mistaken for an empty list.
            $tweets = json_decode($json, false, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException $error) {
            throw new InvalidArchiveException($error->getMessage(), previous: $error);
        }
        if (! \is_array($tweets)) {
            throw new InvalidArchiveException('expected an array of tweets');
        }
        $sightings = [];
        foreach ($tweets as $index => $tweet) {
            // A missing media means none, but a null one is a broken tweet.
            $media = $tweet instanceof stdClass && property_exists($tweet, 'media') ? $tweet->media : '';
            if (! $tweet instanceof stdClass
                || ! \is_string($tweet->full_text ?? null)
                || ! \is_string($tweet->way ?? null)
                || ! \is_string($tweet->created_at ?? null)
                || ! \is_string($media)) {
                throw new InvalidArchiveException("tweet $index is not a tweet");
            }
            $time = self::twitterTime($tweet->created_at);
            if ($time !== null && Tracker::isWay($tweet->way)) {
                $sightings[] = Tracker::sighting($tweet->way, $time, $media === '' ? null : $media, $tweet->full_text);
            }
        }
        usort($sightings, static fn (array $left, array $right): int => $left['time'] <=> $right['time']);

        return $sightings;
    }

    /**
     * Twitter's own format, `Sun Nov 13 10:47:15 +0000 2011`.
     */
    public static function twitterTime(string $text): ?int
    {
        if (! preg_match('/^(\w{3}) (\w{3}) (\d{1,2}) (\d{1,2}):(\d{1,2}):(\d{1,2}) ([+-])(\d{2}):?(\d{2}) (\d{4})$/', $text, $match)) {
            return null;
        }
        [, $weekday, $monthName, $day, $hour, $minute, $second, $sign, $offsetHours, $offsetMinutes, $year] = $match;
        $month = self::MONTHS[$monthName] ?? null;
        [$day, $hour, $minute, $second, $offsetHours, $offsetMinutes, $year]
            = array_map(intval(...), [$day, $hour, $minute, $second, $offsetHours, $offsetMinutes, $year]);
        if ($month === null || $hour > 23 || $minute > 59 || $second > 59 || $offsetMinutes > 59 || ! checkdate($month, $day, $year)) {
            return null;
        }
        $midnight = DateTimeImmutable::createFromFormat('!Y-n-j', \sprintf('%04d-%d-%d', $year, $month, $day), new DateTimeZone('UTC'));
        if ($midnight === false || self::WEEKDAYS[(int) $midnight->format('w')] !== $weekday) {
            return null;
        }
        $offset = ($sign === '-' ? -1 : 1) * ($offsetHours * 3600 + $offsetMinutes * 60);

        return $midnight->getTimestamp() + $hour * 3600 + $minute * 60 + $second - $offset;
    }

    /**
     * Pairs every `in` with the last `out` before it. Two `out` events in a row
     * overwrite each other, because they mean a return went missing.
     *
     * @param  list<array{way: string, time: int}>  $sightings
     * @return array<string, mixed>
     */
    public static function stats(array $sightings): array
    {
        $stats = [
            'total' => \count($sightings),
            'ins' => 0,
            'outs' => 0,
            'first' => $sightings[0]['time'] ?? null,
            'last' => $sightings === [] ? null : $sightings[array_key_last($sightings)]['time'],
            'outings' => 0,
            'unpaired' => 0,
            'avg_outing_secs' => 0,
            'median_outing_secs' => 0,
            'longest' => null,
            'shortest' => null,
            'by_hour_out' => array_fill(0, 24, 0),
            'by_hour_in' => array_fill(0, 24, 0),
            'by_weekday_out' => array_fill(0, 7, 0),
        ];
        $pending = null;
        $durations = [];

        foreach ($sightings as ['way' => $way, 'time' => $time]) {
            $hour = (int) gmdate('G', $time);
            if ($way === 'out') {
                $stats['outs']++;
                $stats['by_hour_out'][$hour]++;
                $stats['by_weekday_out'][(int) gmdate('w', $time)]++;
                if ($pending !== null) {
                    $stats['unpaired']++;
                }
                $pending = $time;

                continue;
            }
            $stats['ins']++;
            $stats['by_hour_in'][$hour]++;
            if ($pending === null) {
                continue;
            }
            $outing = ['out_at' => $pending, 'in_at' => $time, 'secs' => $time - $pending];
            $pending = null;
            $durations[] = $outing['secs'];
            if ($stats['longest'] === null || $outing['secs'] > $stats['longest']['secs']) {
                $stats['longest'] = $outing;
            }
            if ($stats['shortest'] === null || $outing['secs'] < $stats['shortest']['secs']) {
                $stats['shortest'] = $outing;
            }
        }
        if ($pending !== null) {
            $stats['unpaired']++;
        }
        $stats['outings'] = \count($durations);
        if ($durations !== []) {
            $stats['avg_outing_secs'] = intdiv(array_sum($durations), \count($durations));
            sort($durations);
            $stats['median_outing_secs'] = $durations[intdiv(\count($durations), 2)];
        }

        return $stats;
    }
}
